# Weather Service IPM Interface

In [2]:
from agroservices.ipm.ipm import IPM
from agroservices.ipm.datadir import datadir
ipm = IPM()

## MetaDataServices

* Get IPM Platform weather parameters and their description (id, name, description and units)

In [3]:
ipm.get_parameter()

[{'id': 1001,
  'name': 'Instantaneous temperature at 2m',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'AVG'},
 {'id': 1002,
  'name': 'Mean air temperature at 2m',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'AVG'},
 {'id': 1003,
  'name': 'Minimum air temperature at 2m',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'MIN'},
 {'id': 1004,
  'name': 'Maximum air temperature at 2m',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'MAX'},
 {'id': 1021,
  'name': 'Instantaneous temperature in canopy',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'AVG'},
 {'id': 1022,
  'name': 'Mean air temperature in canopy',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'AVG'},
 {'id': 1023,
  'name': 'Minimum air temperature in canopy',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'MIN'},
 {'id': 1024,
  'name': 'Maximum air temperature in canopy',
  'description': No

* Get the controle quality informations (id, name and description)

In [3]:
ipm.get_qc()

[{'id': 0,
  'name': 'No quality control performed',
  'description': 'There has been no quality control of the parameter'},
 {'id': 1,
  'name': 'OK. Parameter comes from a qualified external source',
  'description': 'The external source has its own quality controls, so we can trust it'},
 {'id': 2,
  'name': 'OK. Quality control performed by IPM Decisions',
  'description': "The parameter has been QC'd by the IPM Decisions platform and found to be OK"},
 {'id': 4,
  'name': 'Failed. Not an actual number',
  'description': 'The value is not a valid number'},
 {'id': 8, 'name': 'Failed. Interval test', 'description': 'TODO'},
 {'id': 16, 'name': 'Failed. Logical test', 'description': 'TODO'},
 {'id': 32, 'name': 'Failed. Step test', 'description': 'TODO'},
 {'id': 64, 'name': 'Failed. Freeze test', 'description': 'TODO'},
 {'id': 128,
  'name': 'Failed. Missing value(s)',
  'description': 'Either empty or NULL values have been detected'}]

* Get a schema that describes the IPM Decision platform's format for exchange of weather data

In [4]:
ipm.get_schema_weatherdata() 

{'$schema': 'http://json-schema.org/draft-04/schema#',
 'title': 'Weather Data',
 'type': 'object',
 'additionalProperties': False,
 'description': 'Version 0.1. The schema describes the weather data format for the IPM Decisions platform. See an example here: https://platform.ipmdecisions.net/api/wx/rest/weatheradapter/yr/?longitude=14.3711&latitude=67.2828&altitude=70',
 '$id': 'https://platform.ipmdecisions.net/api/wx/rest/schema/weatherdata',
 'properties': {'timeStart': {'type': 'string',
   'format': 'date-time',
   'description': 'The timestamp of the first weather observation. Format: "yyyy-MM-dd\'T\'HH:mm:ssXXX", e.g. 2020-04-09T18:00:00+02:00',
   'title': "Time start (yyyy-MM-dd'T'HH:mm:ssXXX)"},
  'timeEnd': {'type': 'string',
   'format': 'date-time',
   'description': 'The timestamp of the last weather observation. Format: "yyyy-MM-dd\'T\'HH:mm:ssXXX", e.g. 2020-04-09T18:00:00+02:00',
   'title': "Time end (yyyy-MM-dd'T'HH:mm:ssXXX)"},
  'interval': {'type': 'integer',
   

* Validates the posted weather data against the Json schema

In [5]:
ipm.post_schema_weatherdata_validate(jsonfile=datadir + 'weather_data.json')

{'isValid': True}

## Weather Adapter Services

* Get a list of WeatherAdapterService available on ipm plateform

In [6]:
ws = ipm.get_weatherdatasource()
{k:v['description'] for k,v in ws.items()}

{'no.met.locationforecast': '9-day forecasts for the entire planet. 2.5 km resolution in the Nordic-Baltic region, 9km elsewhere',
 'ie.gov.data': '9-day forecasts for Ireland.',
 'dk.dmi.pointweather': 'Seasonal weather data and forecasts for Denmark. Grid based',
 'se.slu.lantmet': 'Seasonal weather data and forecasts for Sweden. Grid based',
 'net.ipmdecisions.dwd.forecast.germany': "27 hour weather forecasts for Germany and surrounding areas. 2.5 km resolution. Pulls data from DWD's Open Data server (https://www.dwd.de/EN/ourservices/opendata/opendata.html). Made available by a custom adapter created by Met Norway and NIBIO.",
 'net.ipmdecisions.dwd.forecast.europe': "30 hour weather forecasts for EU. 7 km resolution. Pulls data from DWD's Open Data server (https://www.dwd.de/EN/ourservices/opendata/opendata.html). Made available by a custom adapter created by Met Norway and NIBIO.",
 'net.ipmdecisions.dwd.euroweather': "Seasonal data and 72 hour weather forecasts for EU. 7 km reso

* Get weather observations for one weatheradapter service

In this example we choose Finnish Meteorological Institute measured data using endpoint='/weatheradapter/fmi/'

This weather resource does not need credentials, so credentials is None    
ignone= True    
interval=3600    
parameters = [1002, 3002] to select weather parameters 'Mean air temperature at 2m' and 'Mean RH at 2m' see ipm.get_parameters() function for description    
timeStart='2020-06-12T00:00:00+03:00'    
timeEnd='2020-07-03T00:00:00+03:00'    
weatherStationId=101104

In [7]:
source = ipm.get_weatherdatasource('fi.fmi.observation.station')
ipm.get_weatheradapter(source)

{'timeStart': '2020-01-01T23:00:00Z',
 'timeEnd': '2020-01-02T00:00:00Z',
 'interval': 3600,
 'weatherParameters': [2001, 1002, 3002],
 'locationWeatherData': [{'longitude': 28.29911,
   'latitude': 68.84928,
   'altitude': 0.0,
   'amalgamation': [0, 0, 0],
   'data': [[0.6, -7.5, 94.0], [0.3, -7.4, 93.0]],
   'length': 2,
   'qc': [1, 1, 1],
   'width': 3}]}

<div class="alert alert-warning">

 Among Weather resource forecast resources are available. When you choose forecast resource thank to use get_weatheradappter_forecast

<div>

In [9]:
source = ipm.get_weatherdatasource('no.met.locationforecast')
ipm.get_weatheradapter(source)


{'timeStart': '2023-05-22T23:00:00Z',
 'timeEnd': '2023-06-01T12:00:00Z',
 'interval': 3600,
 'weatherParameters': [1001, 3001, 2001, 4002],
 'locationWeatherData': [{'longitude': 110.38587558393341,
   'latitude': 84.32042558195184,
   'altitude': 0.0,
   'amalgamation': [0, 0, 0, 0],
   'data': [[-9.2, 84.7, 0.0, 7.8],
    [-9.9, 84.7, 0.0, 7.8],
    [-9.9, 84.1, 0.0, 7.4],
    [-9.8, 82.7, 0.0, 7.1],
    [-9.6, 81.3, 0.0, 6.8],
    [-9.4, 80.7, 0.0, 6.4],
    [-9.1, 79.1, 0.0, 6.1],
    [-8.9, 77.3, 0.0, 5.6],
    [-8.8, 76.1, 0.0, 5.2],
    [-8.8, 75.9, 0.0, 4.9],
    [-8.8, 74.6, 0.0, 4.5],
    [-8.8, 74.4, 0.0, 3.9],
    [-8.8, 74.2, 0.0, 3.2],
    [-8.9, 73.7, 0.0, 2.4],
    [-9.0, 73.5, 0.0, 1.8],
    [-9.1, 73.6, 0.0, 1.4],
    [-9.2, 73.6, 0.0, 1.5],
    [-9.2, 75.0, 0.0, 2.1],
    [-9.2, 76.1, 0.0, 2.6],
    [-9.1, 77.0, 0.0, 3.1],
    [-9.2, 77.3, 0.0, 3.7],
    [-9.3, 78.3, 0.0, 4.1],
    [-9.3, 78.7, 0.0, 4.7],
    [-9.2, 79.0, 0.0, 5.6],
    [-9.3, 80.5, 0.0, 6.4],
    [

## WeatherDataService

* Get a list of all the available weather data sources

In [10]:
ipm.get_weatherdatasource().keys()

dict_keys(['no.met.locationforecast', 'ie.gov.data', 'dk.dmi.pointweather', 'se.slu.lantmet', 'net.ipmdecisions.dwd.forecast.germany', 'net.ipmdecisions.dwd.forecast.europe', 'net.ipmdecisions.dwd.euroweather', 'net.ipmdecisions.meteofrance.forecast', 'fi.fmi.forecast.location', 'fi.fmi.observation.station', 'no.nibio.lmt', 'com.meteobot', 'info.fruitweb', 'net.ipmdecisions.metos', 'fr.meteo-concept.api'])

* Search for weather data sources that serve the specific location. The location can by any valid Geometry, such as Point or Polygon. Example GeoJson input 

In [12]:
ipm.post_weatherdatasource_location(tolerance=0,geoJsonfile=datadir + 'GeoJson.json')

[{'id': 'no.met.locationforecast',
  'name': 'Met Norway Locationforecast',
  'description': '9-day forecasts for the entire planet. 2.5 km resolution in the Nordic-Baltic region, 9km elsewhere',
  'public_URL': 'https://api.met.no/weatherapi/locationforecast/1.9/documentation',
  'endpoint': '{WEATHER_API_URL}/rest/weatheradapter/yr/',
  'authentication_type': 'NONE',
  'needs_data_control': 'false',
  'access_type': 'location',
  'priority': 0,
  'temporal': {'forecast': 9,
   'historic': {'start': None, 'end': None},
   'intervals': [3600]},
  'parameters': {'common': [1001, 3001, 2001, 4002], 'optional': None},
  'spatial': {'countries': None, 'geoJSON': '{"type": "Sphere"}'},
  'organization': {'name': 'The Norwegian Meteorological Institute',
   'country': 'Norway',
   'address': 'Postboks 43 Blindern',
   'postal_code': '0371',
   'city': 'Oslo',
   'email': 'david.melchior@met.no',
   'url': 'https://www.met.no/en'},
  'active': True,
  'endpointFullPath': 'https://platform.ipm

* Search for weather data sources that serve the specific point

In [13]:
ipm.get_weatherdatasource_location_point(
    latitude="59.678835236960765",
    longitude="12.01629638671875",
    tolerance=0)

[{'id': 'no.met.locationforecast',
  'name': 'Met Norway Locationforecast',
  'description': '9-day forecasts for the entire planet. 2.5 km resolution in the Nordic-Baltic region, 9km elsewhere',
  'public_URL': 'https://api.met.no/weatherapi/locationforecast/1.9/documentation',
  'endpoint': '{WEATHER_API_URL}/rest/weatheradapter/yr/',
  'authentication_type': 'NONE',
  'needs_data_control': 'false',
  'access_type': 'location',
  'priority': 0,
  'temporal': {'forecast': 9,
   'historic': {'start': None, 'end': None},
   'intervals': [3600]},
  'parameters': {'common': [1001, 3001, 2001, 4002], 'optional': None},
  'spatial': {'countries': None, 'geoJSON': '{"type": "Sphere"}'},
  'organization': {'name': 'The Norwegian Meteorological Institute',
   'country': 'Norway',
   'address': 'Postboks 43 Blindern',
   'postal_code': '0371',
   'city': 'Oslo',
   'email': 'david.melchior@met.no',
   'url': 'https://www.met.no/en'},
  'active': True,
  'endpointFullPath': 'https://platform.ipm